### PR TITLE
Snap the image offset to pixel grid when no scaling is applied

### LINKED ISSUE #66259: Linux: Image gets rendered with nearest filtering
## Steps to Reproduce

1. Clone and https://github.com/flutter/gallery.
2.
```
flutter config --enable-linux-desktop
flutter create .
flutter run -d linux
```

**Expected results:**

Flutter logo should be rendered with reasonable anti-aliasing.

**Actual results:**

Flutter logo looks weird.

Screenshot from 2020-09-21 17-21-42

Inspection with RenderDoc shows that the image is rendered with `GL_TEXTURE_{MIN,MAG}_FILTER = GL_NEAREST`. Which is not the behavior on other platforms, at least.

<details>
  <summary>Logs</summary>

[run.log](https://github.com/flutter/flutter/files/5255147/run.log)

```
$ flutter analyze

Analyzing gallery...                                                    
No issues found! (ran in 1.8s)
```

```
$ flutter doctor -v    
[✓] Flutter (Channel master, 1.22.0-10.0.pre.269, on Linux, locale en_US.UTF-8)
    • Flutter version 1.22.0-10.0.pre.269 at /home/ishitatsuyuki/Documents/flutter
    • Framework revision 51fa704642 (2 days ago), 2020-09-19 11:27:04 -0700
    • Engine revision 4b8477d115
    • Dart version 2.10.0 (build 2.10.0-142.0.dev)

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/linux#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, please use
      `flutter config --android-sdk` to update to that location.


[✓] Linux toolchain - develop for Linux desktop
    • clang version 10.0.1
    • cmake version 3.18.2
    • ninja version 1.10.1
    • pkg-config version 1.7.3

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/linux#android-setup for detailed instructions).

[✓] Connected device (1 available)
    • Linux (desktop) • linux • linux-x64 • Linux

! Doctor found issues in 2 categories.
```

OS: Arch Linux, Flutter set up from git master
Graphics: Mesa 20.1.8, AMDGPU, Skia using GLES backend
</details>

### PR DESCRIPTION
**Pull request by ishitatsuyuki**
## Description

See #66259 for an description, and an example of how it might look when the implementation is broken. Apart from fixing the broken cases, this change is intended to not result in any rendering difference.

## Related Issues

Close #66259

## Tests

It might be possible to test this behavior using snapshots (screenshots), but I don't think they are used for this repo yet.

image_rtl_test, which uses hardcoded expected values, was updated to use rounded values, adapting to the behavior of this change. Not the perfect thing to do, but I think it's OK due to the nature of snapshot-like tests.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/painting/decoration_image.dart [hunk 1]
```diff
@@ -561,6 +559,14 @@ void paintImage({
     final Rect sourceRect = alignment.inscribe(
       sourceSize, Offset.zero & inputSize,
     );
+    if (sourceSize == destinationSize) {
+      // If we are rendering the image without scaling, we can improve the
+      // rendering clarity by snapping the image to pixel grid.
```
[the remark this answers is not in the document]
Comment (ishitatsuyuki) [reply packages/flutter/lib/src/painting/decoration_image.dart:564]: I don't think we're using logical pixels here; `sourceSize` already has `scale` multiplied in the code above.

As for canvas transformations, I think it's a legit concern; but it seems that there's no way to determine what kind of transform is applied from the `Canvas` class. Maybe we could grab that inside `RenderImage` using `getTransformTo`, then pass it through a boolean flag to `paintImage`? Does this sound good to you?

**Comment by ishitatsuyuki** [comment]
What are your thoughts on https://github.com/flutter/flutter/pull/67041/files#r498579830? @goderbauer @dnfield

File: packages/flutter/lib/src/painting/decoration_image.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (dnfield) [reply packages/flutter/lib/src/painting/decoration_image.dart:564]: Scale is not the same as the DPR. This is still in logical pixels.

[the remark this answers is not in the document]
Comment (dnfield) [reply packages/flutter/lib/src/painting/decoration_image.dart:564]: Errr, sorry, let me rephrase that - scale is not _necessarily_ the same as the DPR, it's something that might be the DPR but probably is not.

**Review by dnfield — COMMENTED** [review]
We do, in fact, have golden testing in this repo. Unfortunately, it's a little tough to do that with real images right now.

However, internally, we have many many golden tests using images, and this is very likely to break them. I'm also not convinced it's correct given that it's not necessarily putting us on a physical pixel boundary and that it's very likely missing the true logical to physical transformation needed, especially if there are other transforms on the canvas (which will frequently be present at least for the DPR).

**Comment by ishitatsuyuki** [comment]
I think I'm going to close this for now. Though, the silent "nearest" behavior has been broken from the beginning; I think we might want to remove that behavior and just use linear filtering for all cases to ensure correctness.

In the future, I might consider putting this logic into Skia since it's aware of the transforms.
